scoring/score.py

```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
+                f"must be a single pallet from "
+                f"{join_or(repr(x) for x in ZONE_COLOURS)}.\n"
+                f"{bad_highest!r}",
+                code='invalid_highest_pallet',
+            )
+
+        bad_pallets = {}
+        for name, district in self._districts.items():
+            extra = district['pallet_counts'].keys() - ZONE_COLOURS
+            if extra:
+                bad_pallets[name] = extra
+        if bad_pallets:
+            raise InvalidScoresheetException(
+                f"Invalid pallets specified in some districts -- must be from "
+                f"{join_or(repr(x) for x in ZONE_COLOURS)}.\n"
+                f"{bad_pallets!r}",
+                code='invalid_pallets',
+            )
+
+        bad_highest2 = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            if highest and highest not in district['pallet_counts']:
+                bad_highest2[name] = (highest, district['pallet_counts'].keys())
+        if bad_highest2:
+            detail = "\n".join(
+                (
+                    (
+                        f"District {name} has only {join_and(pallets)} so "
+                        f"{highest!r} cannot be the highest."
+                    )
+                    if pallets
+                    else (
+                        f"District {name} has no pallets so {highest!r} cannot "
+                        "be the highest."
+                    )
+                )
+                for name, (highest, pallets) in bad_highest2.items()
+            )
+            raise InvalidScoresheetException(
+                f"Impossible pallets specified as the highest in some districts "
+                f"-- must be a pallet which is present in the district.\n"
+                f"{detail}",
+                code='impossible_highest_pallet',
+            )
+
+        missing_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            pallet_counts = district['pallet_counts']
+            if len(pallet_counts) == 1 and not highest:
+                pallet, = pallet_counts.keys()
+                missing_highest[name] = pallet
+        if missing_highest:
+            raise InvalidScoresheetException(
+                f"Some districts with pallets from a single team are missing "
+                "specification of the highest.\n"
+                f"{missing_highest!r}",
+                code='missing_highest_pallet',
+            )
+
+        totals = collections.Counter()
+        for district in self._districts.values():
+            totals += district['pallet_counts']
+        if any(x > TOKENS_PER_ZONE for x in totals.values()):
+            raise InvalidScoresheetException(
+                f"Too many pallets of some kinds specified, must be no more "
+                f"than {TOKENS_PER_ZONE} of each type.\n"
+                f"{totals!r}",
```

Comment:
Could we say **which** pallet has been listed too many times here?

Reply:
I'd hoped that that would be sufficiently obvious from the totals. The error message here looks like:
> Too many pallets of some kinds specified, must be no more than 6 of each type.
> {'P': 7, 'G': 3}

I guess we could say like:
> Too many 'P' pallets specified, must be no more than 6 of each type.
> Totals: {'P': 7, 'G': 3}

Is that what you had in mind?

Reply:
9c0b0b2 does that.

Reply:
I like this change